sclock.1
.\"     $Id: sclock.1,v 1.2 2007/05/22 14:22:58 pyr Exp $
.\"
.\" Copyright (c) 2007 Pierre-Yves Ritschard <pyr@spootnik.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd May 22, 2007
.Dt SCLOCK 1
.Os
.Sh NAME
.Nm sclock
.Nd simple digital clock
.Sh SYNOPSIS
.Op Fl twelve
.Op Fl apm
.Op Fl blink
.Op Fl blevel Ar percent
.Op Fl bpause Ar seconds
.Op Fl bscript Ar path
.Op Fl seconds
.Op Fl outline
.Op Fl sep
.Op Fl db
.Op Fl \&dw
.Op Fl dh
.Op Fl fg
.Op Fl bg
.Op Fl cw
.Sh DESCRIPTION
The
.Nm
program display the time in a digital led display.
Seconds can be displayed as well as either 12-hour format time or
military time (24-hour format).
.Pp
.Sh OPTIONS
.Nm
accepts all of the standard X Toolkit command line options along with the 
additional options listed below:
.Bl -tag -width Ds
.It Fl twelve
This option indicates that time should be displayed in 12-hour format
followed by the mention
.Em am
or
.Em pm .
The default output format is military time.
.It Fl apm
Display battery status in percents or AC if external power is connected.
Works only on architectures that provide an
.Xr apm 4
interface.
.It Fl blink
Make the column displayed between hours and minutes and - if applicable -
between minutes and seconds blink every second.
The default is not to blink.
.It Fl blevel Ar percent
If APM is supported and enabled this option specifies battery discharge
threshold triggering the
.Fl beepscript .
Default value is 10%.
.It Fl bpause Ar seconds
If beep script is configured this option specifies interval in seconds
between two consecutive beeps.
Default value is 60 seconds.
.It Fl bscript Ar path
A program that will be executed when battery discharge level reaches the
.Ar beeplevel
threshold.
.It Fl seconds
Display seconds.
The default is to only display hours and minutes.
.It Fl outline
Draw a one-pixel line inside the window using the standard foreground color.
No outline is drawn by default.
.It Fl sep
This parameter is used to mimic a wrist watch, it indicates how many pixels
to skip when drawing lines to represent digits.
By default 2 pixels are skipped.
.It Fl db
Separate digits and columns from each other by 
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
By default 2 pixel separate each digit.
.It Fl \&dw
Width of a digit.
The default value is 6 pixel.
.It Fl dh
Height of a digit.
The default value is 10 pixel.
.It Fl cw
Width of the dotted column drawn between elmements of the time.
The default value is 1 pixel.
.El
.Sh RESOURCES
This program uses the 
.I SClock
class, 
widget.  It understands all of the core resource names and classes as well as:
.Pp
.Bl -tag -width Ds -compact
.It Xo
.Ic twelve (class Twelve) 
.Xc
same as the
.Em -twelve
command line option.
.It Xo
.Ic apm (class Apm)
.Xc
same as the
.Em -apm
command line option.
.It Xo
.Ic blink (class Blink) 
.Xc
same as the
.Em -blink
command line option.
.It Xo
.Ic beeplevel (class BeepLevel) 
.Xc
same as the
.Em -blevel
command line option.
.It Xo
.Ic beeppause (class BeepPause) 
.Xc
same as the
.Em -bpause
command line option.
.It Xo
.Ic beepscript (class BeepScript) 
.Xc
same as the
.Em -bscript
command line option.
.It Xo
.Ic seconds (class Seconds) 
.Xc
same as the
.Em -seconds
command line option.
.It Xo
.Ic outline (class Outline) 
.Xc
same as the
.Em -outline
command line option.
.It Xo
.Ic digit.sep (class Digit.Sep) 
.Xc
same as the
.Em -sep
command line option.
.It Xo
.Ic digit.border (class Digit.Border) 
.Xc
same as the
.Em -db
command line option.
.It Xo
.Ic digit.width (class Digit.Width) 
.Xc
same as the
.Em -dw
command line option.
.It Xo
.Ic digit.height (class Digit.Height) 
.Xc
same as the
.Em -dh
command line option.
.It Xo
.Ic column.width (class Column.Width)
.Xc
same as the
.Em -cw
command line option.
.El
.Sh FILES
.Bl -tag -width "/etc/X11/app-defaults/SClockXX" -compact
.It /etc/X11/app-defaults/SClock
specifies required resources
.El
.Sh "SEE ALSO"
.Xr X 7 ,
.Xr xrdb 1 ,
.Xr xclock 1 ,
.Xr time 3C
.Sh AUTHORS
Pierre-Yves Ritschard wrote the program.
Mike Belopuhov has added APM support and beeps.


sclock.c
/* $Id: sclock.c,v 1.2 2007/05/22 14:30:11 pyr Exp $ */
/*
 * Copyright (c) 2007 Pierre-Yves Ritschard <pyr@spootnik.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/param.h>
#include <err.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <poll.h>
#include <X11/Xlib.h>
#include <X11/Xresource.h>
#include <X11/keysym.h>

#if defined(__amd64__) || defined(__i386__) || defined(__macppc__)
#include <sys/ioctl.h>
#include <machine/apmvar.h>
#include <fcntl.h>
#define HAVE_APM	1
#endif

#ifndef APPLOADDIR
#define APPLOADDIR	"/etc/X11/app-defaults/"
#endif
#define	CLASS_NAME	"SClock"

#define DEFAULT_GEOM	"80x15"
#define DEFAULT_FG	"green"
#define DEFAULT_BG	"#2d2d2d"
#define DEFAULT_FONT	"fixed"
#define DIGIT_BORDER	2
#define DIGIT_SEP	1
#define DIGIT_WIDTH	6
#define DIGIT_HEIGHT	10
#define COL_WIDTH	1

#define BH_TOP	0x01
#define BH_MID	0x02
#define BH_BOT	0x04
#define BV_TLF	0x08
#define BV_TRT	0x10
#define BV_BLF	0x20
#define BV_BRT	0x40

#define D0		BH_TOP|BH_BOT|BV_TLF|BV_TRT|BV_BLF|BV_BRT
#define	D1		BV_TRT|BV_BRT
#define D2		BH_TOP|BH_MID|BH_BOT|BV_TRT|BV_BLF
#define D3		BH_TOP|BH_MID|BH_BOT|BV_TRT|BV_BRT
#define D4		BH_MID|BV_TLF|BV_TRT|BV_BRT
#define D5		BH_TOP|BH_MID|BH_BOT|BV_TLF|BV_BRT
#define D6		BH_TOP|BH_MID|BH_BOT|BV_TLF|BV_BLF|BV_BRT
#define	D7		BH_TOP|BV_TRT|BV_BRT
#define D8		BH_TOP|BH_MID|BH_BOT|BV_TLF|BV_TRT|BV_BLF|BV_BRT
#define D9		BH_TOP|BH_MID|BH_BOT|BV_TLF|BV_TRT|BV_BRT
#define DA		BH_TOP|BH_MID|BV_TLF|BV_TRT|BV_BLF|BV_BRT
#define DC		BH_TOP|BH_BOT|BV_TLF|BV_BLF

struct area {
	int		 x;
	int		 y;
	int		 w;
	int		 h;
};

struct widget {
	int			 flags;
#define F_TWELVE	 	 0x01
#define F_BLINK		 	 0x02
#define F_SECONDS		 0x04
#define F_OUTLINE		 0x08
#define F_SHADE			 0x10
#define F_APM			 0x20
	int			 screen;
	Display			*dpy;
	Window			 root;
	Window			 win;
	XColor			 fg;
	XColor			 bg;
	XWindowAttributes	 wa;
	XFontStruct		*font;
	GC			 gc;
	struct area		 geom;
#define NAREAS			 11
	struct area		 areas[NAREAS];
	time_t			 before;
#ifdef HAVE_APM
	int			 apmdev;
	int			 beeplevel;
	int			 beeppause;
	char			 beepscript[PATH_MAX];
#endif
	int			 sep;
	int			 db;
	int			 dw;
	int			 dh;
	int			 cw;
	int			 xofft;
	int			 yofft;
};

void	resource_init(struct widget *, int, char *[]);
void	widget_init(struct widget *);
void	widget_setup(struct widget *);
void	widget_calcsize(struct widget *, int, int, int, int);
void	widget_draw(struct widget *, int);
void	widget_draw_column(struct widget *, int);
void	widget_draw_digit(struct widget *, int, int);
void	widget_event(struct widget *, XEvent *);

void
usage(void)
{
	extern const char	*__progname;

	fprintf(stderr, "usage: %s", __progname);
	exit(1);
}

void
widget_init(struct widget *w)
{
	if ((w->dpy = XOpenDisplay(NULL)) == NULL)
		errx(1, "cannot open display");
	w->screen = DefaultScreen(w->dpy);
	w->root = RootWindow(w->dpy, w->screen);
	XGetWindowAttributes(w->dpy, w->root, &w->wa);
}

void
widget_areas(struct widget *w)
{
	int			 width;
	struct area		*last_area;

	width = (w->dw + w->db) * 4 + w->cw;

	if (w->flags & F_SECONDS)
		width += w->db + (w->dw + w->db) * 2 + w->cw;

	if (w->flags & F_TWELVE)
		width += w->db + MAX(XTextWidth(w->font, "am", 2),
		    XTextWidth(w->font, "pm", 2));

	if (width > w->geom.w)
		errx(1, "invalid geometry");

#ifdef HAVE_APM
	if (w->flags & F_APM)
		w->xofft = 0;
	else
#endif
		w->xofft = (w->geom.w / 2) - (width / 2);

	w->yofft = (w->geom.h / 2) - (w->dh / 2);

	w->areas[0].x = w->xofft;
	w->areas[0].y = w->yofft;
	w->areas[0].w = w->dw;
	w->areas[0].h = w->dh;

	w->areas[1].x = w->areas[0].x + w->areas[0].w + w->db;
	w->areas[1].y = w->yofft;
	w->areas[1].w = w->dw;
	w->areas[1].h = w->dh;

	w->areas[2].x = w->areas[1].x + w->areas[1].w + w->db;
	w->areas[2].y = w->yofft;
	w->areas[2].w = w->cw;
	w->areas[2].h = w->dh;

	w->areas[3].x = w->areas[2].x + w->areas[2].w + w->db;
	w->areas[3].y = w->yofft;
	w->areas[3].w = w->dw;
	w->areas[3].h = w->dh;

	w->areas[4].x = w->areas[3].x + w->areas[3].w + w->db;
	w->areas[4].y = w->yofft;
	w->areas[4].w = w->dw;
	w->areas[4].h = w->dh;

	last_area = &w->areas[4];

	if (w->flags & F_SECONDS) {
		w->areas[5].x = w->areas[4].x + w->areas[4].w + w->db;
		w->areas[5].y = w->yofft;
		w->areas[5].w = w->cw;
		w->areas[5].h = w->dh;

		w->areas[6].x = w->areas[5].x + w->areas[5].w + w->db;
		w->areas[6].y = w->yofft;
		w->areas[6].w = w->dw;
		w->areas[6].h = w->dh;

		w->areas[7].x = w->areas[6].x + w->areas[6].w + w->db;
		w->areas[7].y = w->yofft;
		w->areas[7].w = w->dw;
		w->areas[7].h = w->dh;

		last_area = &w->areas[7];
	}

	if (w->flags & F_TWELVE) {
		w->areas[8].x = last_area->x + last_area->w + w->db;
		w->areas[8].y = w->yofft + w->dh - w->font->descent;
	}

#ifdef HAVE_APM
	if (w->flags & F_APM) {
		/* right aligned */
		w->areas[10].x = w->geom.w - w->db * 5;
		w->areas[10].y = w->yofft;
		w->areas[10].w = w->dw;
		w->areas[10].h = w->dh;

		w->areas[9].x = w->areas[10].x - w->areas[10].w - w->db;
		w->areas[9].y = w->yofft;
		w->areas[9].w = w->dw;
		w->areas[9].h = w->dh;
	}
#endif
}

void
widget_setup(struct widget *w)
{
	XSetWindowAttributes	swa;
	XGCValues		gcv;

	bzero(&swa, sizeof(swa));
	bzero(&gcv, sizeof(gcv));
	w->win = XCreateWindow(w->dpy, w->root, w->geom.x, w->geom.y,
	    w->geom.w, w->geom.h, 0,
	    CopyFromParent, CopyFromParent, CopyFromParent, 0, &swa);
	XStoreName(w->dpy, w->win, "sclock");

	widget_areas(w);

	w->gc = XCreateGC(w->dpy, w->win, 0, &gcv);
	XSetForeground(w->dpy, w->gc, w->fg.pixel);
	XSetBackground(w->dpy, w->gc, w->bg.pixel);
	XSetFont(w->dpy, w->gc, w->font->fid);

	XSelectInput(w->dpy, w->win,
	    ExposureMask | StructureNotifyMask);
	XSetWindowBackground(w->dpy, w->win, w->bg.pixel);
	XSetWindowBackground(w->dpy, w->win, w->bg.pixel);
	XMapWindow(w->dpy, w->win);
}

void
widget_draw(struct widget *w, int forced)
{
	time_t		now;
	struct tm	*tm;
	int		hour;
	int		am;
	XTextItem	ti;

	now = time(NULL);
	if (now == w->before && !forced)
		return;
	w->before = now;
	XClearWindow(w->dpy, w->win);

	if (w->flags & F_OUTLINE) {
		XDrawLine(w->dpy, w->win, w->gc, 0, 0, 0, w->geom.h - 1);
		XDrawLine(w->dpy, w->win, w->gc, 0, 0, w->geom.w - 1, 0);
		XDrawLine(w->dpy, w->win, w->gc, w->geom.w - 1, 0,
		    w->geom.w - 1, w->geom.h - 1);
		XDrawLine(w->dpy, w->win, w->gc, 0, w->geom.h - 1,
		    w->geom.w - 1, w->geom.h - 1);
	}

	now = time(NULL);
	tm = localtime(&now);
	w->before = now;

	hour = tm->tm_hour;
	if (hour > 12)
		am = 0;
	else
		am = 1;
	if (w->flags & F_TWELVE)
		hour = tm->tm_hour % 12;

	widget_draw_digit(w, 0, hour / 10);
	widget_draw_digit(w, 1, hour % 10);
	if (!(w->flags & F_BLINK) || now % 2 == 0)
		widget_draw_column(w, 2);
	widget_draw_digit(w, 3, tm->tm_min / 10);
	widget_draw_digit(w, 4, tm->tm_min % 10);
	if (w->flags & F_SECONDS) {
		if (!(w->flags & F_BLINK) || now % 2 == 0)
			widget_draw_column(w, 5);
		widget_draw_digit(w, 6, tm->tm_sec / 10);
		widget_draw_digit(w, 7, tm->tm_sec % 10);
	}
	if (w->flags & F_TWELVE) {
		bzero(&ti, sizeof(ti));
		ti.chars = (am)?"am":"pm";
		ti.nchars = 2;
		ti.font = w->font->fid;
		XDrawText(w->dpy, w->win, w->gc,
		    w->areas[8].x, w->areas[8].y, &ti, 1);
	}
#ifdef HAVE_APM
	if (w->flags & F_APM) {
		struct apm_power_info	api;
		static time_t		last;

		if (ioctl(w->apmdev, APM_IOC_GETPOWER, &api) == -1)
			err(1, "APM_IOC_GETPOWER");

		if (api.ac_state == APM_AC_ON) {
			widget_draw_digit(w, 9, 10);  /* A */
			widget_draw_digit(w, 10, 11); /* C */
		} else {
			if (api.battery_life == 100)
				api.battery_life--;
			widget_draw_digit(w, 9, api.battery_life / 10);
			widget_draw_digit(w, 10, api.battery_life % 10);

			if (w->beeplevel > 0 && strlen(w->beepscript) > 0 &&
			    w->beeplevel > api.battery_life &&
			    (last == 0 || now - last > w->beeppause)) {
				system(w->beepscript);
				last = now;
			}
		}
	}
#endif
	XFlush(w->dpy);
	XSync(w->dpy, 0);
}

void
widget_draw_column(struct widget *w, int a_idx)
{
	struct area	*a;

	a = &w->areas[a_idx];
	
	XFillRectangle(w->dpy, w->win, w->gc,
	    a->x, a->y + (a->h / 2) - (w->cw * 2), a->w, w->cw);
	XFillRectangle(w->dpy, w->win, w->gc,
	    a->x, a->y + (a->h / 2) + w->cw, a->w, w->cw);
}

void
widget_draw_digit(struct widget *w, int a_idx, int digit)
{
	int		 mask;
	struct area	*a;
	static int	 digits[] = { D0, D1, D2, D3, D4, D5, D6, D7, D8, D9,
				      DA, DC };

	a = &w->areas[a_idx];
	mask = digits[digit];

	/* top */
	if (mask & BH_TOP)
		XDrawLine(w->dpy, w->win, w->gc, a->x + w->sep, a->y,
		    a->x + a->w - (w->sep * 2), a->y);

	/* top vert left */
	if (mask & BV_TLF)
		XDrawLine(w->dpy, w->win, w->gc, a->x, a->y + w->sep,
		    a->x, a->y + (a->h / 2) - w->sep);

	/* top vert right */
	if (mask & BV_TRT)
		XDrawLine(w->dpy, w->win, w->gc,
		    a->x + a->w - w->sep, a->y + w->sep,
		    a->x + a->w - w->sep, a->y + (a->h / 2) - w->sep);

	/* middle */
	if (mask & BH_MID)
		XDrawLine(w->dpy, w->win, w->gc, a->x + w->sep,
		    a->y + (a->h / 2), a->x + a->w - (w->sep * 2),
		    a->y + (a->h / 2));

	/* bottom vert left */
	if (mask & BV_BLF)
		XDrawLine(w->dpy, w->win, w->gc, a->x,
		    a->y + (a->h / 2) + w->sep,
		    a->x, a->y + a->h - w->sep);

	/* bottom vert right */
	if (mask & BV_BRT)
		XDrawLine(w->dpy, w->win, w->gc, a->x + a->w - w->sep,
		    a->y + (a->h / 2) + w->sep, a->x + a->w - w->sep,
		    a->y + a->h - w->sep);

	/* bottom */
	if (mask & BH_BOT)
		XDrawLine(w->dpy, w->win, w->gc, a->x + w->sep,
		    a->y + a->h,
		    a->x + a->w - (w->sep * 2), a->y + a->h);
}

void
widget_event(struct widget *w, XEvent *ev)
{
	switch (ev->type) {
	case Expose:
		if (ev->xexpose.count == 0)
			widget_draw(w, 1);
		break;
	case ConfigureNotify:
		w->geom.x = ev->xconfigure.x;
		w->geom.y = ev->xconfigure.y;
		w->geom.w = ev->xconfigure.width;
		w->geom.h = ev->xconfigure.height;
		widget_areas(w);
		break;
	case DestroyNotify:
		XFreeGC(w->dpy, w->gc);
		XDestroyWindow(w->dpy, w->win);
		XCloseDisplay(w->dpy);
		exit(0);
	}
}

void
resource_init(struct widget *w, int argc, char *argv[])
{
	const char		*estr;
	char			*dbstr;
	char			*dummy;
	char			*sarg;
	XrmDatabase		 db;
	XrmDatabase		 filedb;
	XrmValue		 val;
	XColor			 dump;

	static XrmOptionDescRec	 opts[] = {
		{ "-twelve",	"twelve",	XrmoptionNoArg, "on" },
		{ "-blink",	"blink",	XrmoptionNoArg, "on" },
		{ "-seconds",	"seconds",	XrmoptionNoArg, "on" },
		{ "-outline",	"outline",	XrmoptionNoArg, "on" },
#ifdef HAVE_APM
		{ "-apm",	"apm",		XrmoptionNoArg, "on" },
		{ "-blevel",	"beeplevel",	XrmoptionSepArg, NULL },
		{ "-bpause",	"beeppause",	XrmoptionSepArg, NULL },
		{ "-bscript",	"beepscript",	XrmoptionSepArg, NULL },
#endif
		{ "-sep",	"digit.sep",	XrmoptionSepArg, NULL },
		{ "-db",	"digit.border",	XrmoptionSepArg, NULL },
		{ "-dw",	"digit.width",	XrmoptionSepArg, NULL },
		{ "-dh",	"digit.height",	XrmoptionSepArg, NULL },
		{ "-fg",	"foreground",	XrmoptionSepArg, NULL },
		{ "-bg",	"background",	XrmoptionSepArg, NULL },
		{ "-cw",	"column.width",	XrmoptionSepArg, NULL },
		{ "-geometry",	"geometry",	XrmoptionSepArg, NULL }
	};

	XrmInitialize();
	db = NULL;

	if ((dbstr = XResourceManagerString(w->dpy)) != NULL)
		db = XrmGetStringDatabase(dbstr);

	if ((filedb = XrmGetFileDatabase(APPLOADDIR CLASS_NAME)) != NULL)
		XrmMergeDatabases(filedb, &db);

	XrmParseCommand(&db, opts, sizeof(opts) / sizeof(opts[0]), "sclock",
	    &argc, argv);

	if (XrmGetResource(db, "sclock.twelve", "SClock.Twelve", &dummy, &val))
		if (strcmp((char *)val.addr, "on") == 0)
			w->flags |= F_TWELVE;
	if (XrmGetResource(db, "sclock.blink", "SClock.Blink", &dummy, &val))
		if (strcmp((char *)val.addr, "on") == 0)
			w->flags |= F_BLINK;
	if (XrmGetResource(db, "sclock.seconds", "SClock.Seconds",
	    &dummy, &val))
		if (strcmp((char *)val.addr, "on") == 0)
			w->flags |= F_SECONDS;
	if (XrmGetResource(db, "sclock.outline", "SClock.Outline",
	    &dummy, &val))
		if (strcmp((char *)val.addr, "on") == 0)
			w->flags |= F_OUTLINE;
#ifdef HAVE_APM
	if (XrmGetResource(db, "sclock.apm", "SClock.Apm", &dummy, &val))
		if (strcmp((char *)val.addr, "on") == 0)
			w->flags |= F_APM;
	if (XrmGetResource(db, "sclock.beeplevel", "SClock.BeepLevel",
	    &dummy, &val)) {
		w->beeplevel = strtonum((char *)val.addr, 1, 99, &estr);
		if (estr)
			errx(1, "battery level %s is %s", (char *)val.addr,
			    estr);
	} else
		w->beeplevel = 10;
	if (XrmGetResource(db, "sclock.beeppause", "SClock.BeepPause",
	    &dummy, &val)) {
		w->beeppause = strtonum((char *)val.addr, 10, 300, &estr);
		if (estr)
			errx(1, "beep interval %s is %s", (char *)val.addr,
			    estr);
	} else
		w->beeppause = 60;
	if (XrmGetResource(db, "sclock.beepscript", "SClock.BeepScript",
	    &dummy, &val))
		strlcpy(w->beepscript, (char *)val.addr, sizeof(w->beepscript));
#endif

	if (XrmGetResource(db, "sclock.foreground", "SClock.Foreground",
	    &dummy, &val))
		sarg = (char *)val.addr;
	else
		sarg = DEFAULT_FG;
	XAllocNamedColor(w->dpy, w->wa.colormap, sarg, &dump, &w->fg);

	if (XrmGetResource(db, "sclock.background", "SClock.Background",
	    &dummy, &val))
		sarg = (char *)val.addr;
	else
		sarg = DEFAULT_BG;
	XAllocNamedColor(w->dpy, w->wa.colormap, sarg, &dump, &w->bg);

	if (XrmGetResource(db, "sclock.font", "SClock.Font", &dummy, &val))
		sarg = (char *)val.addr;
	else
		sarg = DEFAULT_FONT;
	w->font = XLoadQueryFont(w->dpy, sarg);

	if (XrmGetResource(db, "sclock.geometry", "SClock.Geometry",
	    &dummy, &val))
		sarg = (char *)val.addr;
	else
		sarg = DEFAULT_GEOM;
	XParseGeometry(sarg, &w->geom.x, &w->geom.y,
	    &w->geom.w, &w->geom.h);

	if (XrmGetResource(db, "sclock.digit.width", "SClock.Digit.Width",
	    &dummy, &val)) {
		w->dw = strtonum((char *)val.addr, 6, w->geom.w, &estr);
		if (estr)
			errx(1, "digit width %s is %s", (char *)val.addr, estr);
	} else
		w->dw = DIGIT_WIDTH;

	if (XrmGetResource(db, "sclock.digit.height", "SClock.Digit.Height",
	    &dummy, &val)) {
		w->dh = strtonum((char *)val.addr, 4, w->geom.h, &estr);
		if (estr)
			errx(1, "digit height %s is %s",
			    (char *)val.addr, estr);
	} else
		w->dh = DIGIT_HEIGHT;

	if (XrmGetResource(db, "sclock.digit.sep", "SClock.Digit.Sep",
	    &dummy, &val)) {
		w->sep = strtonum((char *)val.addr, 0, w->geom.h, &estr);
		if (estr)
			errx(1, "digit height %s is %s",
			    (char *)val.addr, estr);
	} else
		w->sep = DIGIT_SEP;

	if (XrmGetResource(db, "sclock.digit.border", "SClock.Digit.Border",
	    &dummy, &val)) {
		w->db = strtonum((char *)val.addr, 1, w->geom.w, &estr);
		if (estr)
			errx(1, "digit border %s is %s",
			    (char *)val.addr, estr);
	} else
		w->db = DIGIT_BORDER;

	if (XrmGetResource(db, "sclock.column.width", "SClock.Column.Width",
	    &dummy, &val)) {
		w->cw = strtonum((char *)val.addr, 1, w->geom.w, &estr);
		if (estr)
			errx(1, "column width %s is %s",
			    (char *)val.addr, estr);
	} else
		w->cw = COL_WIDTH;
}

int
main(int argc, char *argv[])
{
	struct widget		w;
	XEvent			ev;

	bzero(&w, sizeof(w));
	widget_init(&w);
	resource_init(&w, argc, argv);
	widget_setup(&w);

#ifdef HAVE_APM
	if ((w.flags & F_APM) &&
	    (w.apmdev = open("/dev/apm", O_RDONLY, 0)) == -1)
		err(1, "/dev/apm");
#endif

	while (1) {
		while (XPending(w.dpy)) {
			XNextEvent(w.dpy, &ev);
			widget_event(&w, &ev);
		}
		usleep(250);
		widget_draw(&w, 0);
	}
	return (0);
}
